### PR TITLE
8261455: Automatically generate the CDS archive if necessary

### PR DESCRIPTION
**Pull request by yminqi**
This patch is re-coded after serial fixes related: 8275846, 8276787, 8279018 etc.
original description(PR 5997):
`----------------------------------------------------------------------------------`
When shared archive (dynamic archive) failed to map due to damage of the archive file, dump/run jdk version mismatch or non-existence file etc, the new patch will automatically create a new shared archive if -XX:+AutoCreateSharedArchive specified with the name based on SharedArchiveFile.
This is a revised patch based on the old PR: 5077 and after bug 8273152 integrated.
`-----------------------------------------------------------------------------------`

Tests: tier1,tier2,tier3,tier4

Thanks
Yumin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8261455](https://bugs.openjdk.java.net/browse/JDK-8261455): Automatically generate the CDS archive if necessary
 * [JDK-8272331](https://bugs.openjdk.java.net/browse/JDK-8272331): Automatically generate the CDS archive if necessary (**CSR**)

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6920/head:pull/6920` \
`$ git checkout pull/6920`

Update a local copy of the PR: \
`$ git checkout pull/6920` \
`$ git pull https://git.openjdk.java.net/jdk pull/6920/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6920`

View PR using the GUI difftool: \
`$ git pr show -t 6920`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6920.diff">https://git.openjdk.java.net/jdk/pull/6920.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/filemap.cpp [hunk 1]
```diff
@@ -188,6 +195,22 @@ FileMapInfo::~FileMapInfo() {
...
+  if (_file_open) {
+    os::close(_fd);
+  }
+}
+
+// Do preliminary validation on archive. More checks are in initialization.
+bool FileMapInfo::validate_archive() const {
+  if (!os::file_exists(_full_path)) {
+    return false;
+  }
```
Comment (iklam) [inline src/hotspot/share/cds/filemap.cpp:207]: Is the above check necessary? If the file doesn't exist, will `check_archive()` below return false?

File: src/hotspot/share/cds/dynamicArchive.cpp [hunk 2]
```diff
@@ -410,7 +410,8 @@ void DynamicArchive::dump(const char* archive_name, TRAPS) {
 }
 
 bool DynamicArchive::should_dump_at_vm_exit() {
-  return DynamicDumpSharedSpaces && (ArchiveClassesAtExit != nullptr);
+  return (DynamicDumpSharedSpaces && (ArchiveClassesAtExit != nullptr)) ||
+         (AutoCreateSharedArchive && Arguments::GetSharedDynamicArchivePath() != nullptr);
```
Comment (iklam) [inline src/hotspot/share/cds/dynamicArchive.cpp:414]: I think it's better to do this: if `AutoCreateSharedArchive` is true, and you decide that the archive needs to be updated, then:

```
    DynamicDumpSharedSpaces = true;
    ArchiveClassesAtExit = Arguments::GetSharedDynamicArchivePath();
```

That way we don't need to do the check at line 414 anymore.

**Review by iklam — COMMENTED** [review]
Hi Yumin, looks good. I have some preliminary comments.

File: src/hotspot/share/cds/filemap.cpp (hunk 1 above)
Comment (yminqi) [reply src/hotspot/share/cds/filemap.cpp:207]: It is needed here --- when open a file of name is NULL, os::open first does is checking:
if (strlen(path) > MAX_PATH - 1) {
    errno = ENAMETOOLONG;
    return -1;
  }
It will crash on NULL. Removing it will cause many tests fail.  os:file_exists will first check if it is NULL. 
I will add comment for it.

File: src/hotspot/share/runtime/arguments.cpp [hunk 3]
```diff
@@ -3140,6 +3140,17 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
     DynamicDumpSharedSpaces = true;
   }
 
+  if (AutoCreateSharedArchive) {
+    if (SharedArchiveFile == NULL) {
+      log_info(cds)("-XX:+AutoCreateSharedArchive must work with a valid SharedArchiveFile");
```
Comment (iklam) [inline src/hotspot/share/runtime/arguments.cpp:3145]: The log message should be consistent with other existing messages. E.g., 

```
 fatal("-XX:+VerifyHeavyMonitors requires -XX:+UseHeavyMonitors");
```

So we should have:

```
 -XX:+AutoCreateSharedArchive requires -XX:SharedArchiveFile
```

File: src/hotspot/share/runtime/arguments.cpp [hunk 4]
```diff
@@ -3140,6 +3140,17 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
...
+      return JNI_ERR;
+    }
+    if (ArchiveClassesAtExit != NULL) {
+      log_info(cds)("-XX:+AutoCreateSharedArchive does not work with ArchiveClassesAtExit");
```
Comment (iklam) [inline src/hotspot/share/runtime/arguments.cpp:3149]: These should be `log_warning()`. Otherwise the VM will exit without no apparent reason:
```
$ java -XX:+AutoCreateSharedArchive -version
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

File: src/hotspot/share/cds/filemap.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/filemap.cpp:210]: I think we shouldn't be checking a FileMapInfo when its _file_path is NULL. I added the following assert:

```
bool FileMapInfo::validate_archive() const {
+ assert(_full_path != NULL, "huh");
  if (!os::file_exists(_full_path)) {
    return false;
  }
```

and it crashes here (with dynamicArchive/HelloDynamic.java): 

```
 9 FileMapInfo::validate_archive 
10 FileMapInfo::FileMapInfo 
11 DynamicArchiveBuilder::init_header 
12 DynamicArchiveBuilder::doit 
13 VM_PopulateDynamicDumpSharedSpace::doit 
14 VM_Operation::evaluate 
15 VMThread::evaluate_operation 
16 VMThread::inner_execute 
17 VMThread::loop 
...
```

At this point, we are writing the archive, so we shouldn't check the validity of the (output) archive, because we are going to overwrite it. And why is it NULL?
 
By the way, you added FileMapInfo::validate_archive() for doing this inside FileMapInfo::FileMapInfo():

```
  bool valid = validate_archive();
  if (!valid && !_is_static && AutoCreateSharedArchive) {
    // regenerate shared archive at exit
    DynamicDumpSharedSpaces = true;
    ArchiveClassesAtExit = _full_path;
  }
```

But the above block of code doesn't seem necessary, as it's already checked here inside FileMapInfo::initialize():

```
      FileMapInfo::fail_continue("Initialize dynamic archive failed.");
      if (AutoCreateSharedArchive) {
        DynamicDumpSharedSpaces = true;
        ArchiveClassesAtExit = Arguments::GetSharedDynamicArchivePath();
      }
      return false;
```

File: src/hotspot/share/cds/filemap.cpp [hunk 5]
```diff
@@ -1099,6 +1125,12 @@ class FileHeaderHelper {
       return false;
     }
 
+    if (gen_header._version !=  CURRENT_CDS_ARCHIVE_VERSION) {
+      auto warning_continue = FileMapInfo::fail_continue;
+      warning_continue("The shared archive file version %d which is not current version %d",
+                                    gen_header._version, CURRENT_CDS_ARCHIVE_VERSION);
+    }
```
Comment (iklam) [inline src/hotspot/share/cds/filemap.cpp:1132]: There's no need to add `warning_continue`. `FileMapInfo::fail_continue` means:

- when running with -Xshare:auto, print a warning and continue
- when running with -Xshare:on, print an error and exit

All other calls of `FileMapInfo::fail_continue` expects this behavior.

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 6]
```diff
@@ -1020,6 +1024,10 @@ MapArchiveResult MetaspaceShared::map_archives(FileMapInfo* static_mapinfo, File
     assert(static_mapinfo->mapping_end_offset() == dynamic_mapinfo->mapping_base_offset(), "no gap");
   }
 
+  if (AutoCreateSharedArchive && static_mapinfo != NULL && dynamic_mapinfo == NULL) {
+    warning("AutoCreateSharedArchive will be ignored for static archive");
```
Comment (iklam) [inline src/hotspot/share/cds/metaspaceShared.cpp:1028]: To be consistent, we should use `log_warning(cds)`. Also, the message is unclear. Here, `foo.jsa` is a static archive:

```
$ java -XX:+AutoCreateSharedArchive -XX:SharedArchiveFile=foo.jsa -cp HelloWorld.jar HelloWorld
Java HotSpot(TM) 64-Bit Server VM warning: AutoCreateSharedArchive will be ignored for static archive
Hello World
```

I think it's better to print out:

```
AutoCreateSharedArchive is ignored because foo.jsa is a static archive
```

**Review by iklam — CHANGES_REQUESTED** [review]

**Comment by iklam** [comment]
I tried switching between two builds and the VM crashes:

```
$ rm bar.jsa
$ /bld/fastdebug/images/jdk/bin/java -XX:+AutoCreateSharedArchive \
   -XX:SharedArchiveFile=bar.jsa -cp HelloWorld.jar HelloWorld
Java HotSpot(TM) 64-Bit Server VM warning: AutoCreateSharedArchive will be ignored for static archive
Hello World
```

NOTE: the warning is incorrectly printed.


```
$ /bld/slowdebug/images/jdk/bin/java -XX:+AutoCreateSharedArchive \
   -XX:SharedArchiveFile=bar.jsa -cp HelloWorld.jar HelloWorld
Hello World

[0.002s][info][cds] Opening shared archive: bar.jsa
[0.069s][info][cds] Opening shared archive: /bld/fastdebug/images/jdk/lib/server/classes.jsa
[0.069s][info][cds] trying to map /bld/fastdebug/images/jdk/lib/server/classes.jsa
[0.069s][info][cds] Opened archive /bld/fastdebug/images/jdk/lib/server/classes.jsa.
[0.069s][info][cds] _jvm_ident expected: Java HotSpot(TM) 64-Bit Server VM (slowdebug ....
[0.069s][info][cds]              actual: Java HotSpot(TM) 64-Bit Server VM (fastdebug ....
[0.069s][info][cds] UseSharedSpaces: The shared archive file was created by a different version or build of HotSpot
[0.069s][info][cds] UseSharedSpaces: Initialize static archive failed.
[0.069s][info][cds] UseSharedSpaces: Unable to map shared spaces
Hello World
[1.048s][info][cds] Regenerate MethodHandle Holder classes...
[1.048s][info][cds] Nothing to regenerate for holder classes
[1.048s][info][cds] Regenerate MethodHandle Holder classes...done
# To suppress the following error report, specify this argument
# after -XX: or in .hotspotrc:  SuppressErrorAt=/arguments.hpp:647
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/jdk2/sek/open/src/hotspot/share/runtime/arguments.hpp:647), pid=2165765, tid=2165766
#  assert(Arguments::is_dumping_archive()) failed: dump time only
```

I think you can test this by:

- create a static archive s.jsa
- create a dynamic archive d.jsa base on s.jsa
- corrupt the _jvm_ident of s.jsa
- run with -XX:+AutoCreateSharedArchive -XX:SharedArchiveFile=s.jsa:d.jsa ...

**Comment by iklam** [comment]
Another problem: VM crashes when the default base archive cannot be mapped:

```
$ ls -l /bld/fastdebug/images/jdk/lib/server/classes.jsa
-r--r--r-- 1 iklam dba 13815808 Jan  2 23:22 /bld/fastdebug/images/jdk/lib/server/classes.jsa
$ rm -f /bld/fastdebug/images/jdk/lib/server/classes.jsa
$ touch /bld/fastdebug/images/jdk/lib/server/classes.jsa
$ ls -l /bld/fastdebug/images/jdk/lib/server/classes.jsa
-rw-r--r-- 1 iklam dba 0 Jan  2 23:51 /bld/fastdebug/images/jdk/lib/server/classes.jsa
$ rm -f bar.jsa

$ /bld/fastdebug/images/jdk/bin/java -XX:+AutoCreateSharedArchive \
   -XX:SharedArchiveFile=bar.jsa -cp HelloWorld.jar HelloWorld
#  ...
#  assert(ArchiveClassesAtExit != nullptr) failed: sanity
#
```

I am not sure if there's a good way of testing this inside jtreg. Maybe we can follow open/test/hotspot/jtreg/runtime/cds/appcds/MoveJDKTest.java and make a copy of the JDK and then corrupt its default base archive.

File: src/hotspot/share/runtime/arguments.cpp [hunk 7]
```diff
@@ -3564,9 +3564,9 @@ void Arguments::init_shared_archive_paths() {
           // If +AutoCreateSharedArchive and the specified shared archive does not exist,
           // regenerate the dynamic archive base on default archive.
           if (AutoCreateSharedArchive && !os::file_exists(SharedArchiveFile)) {
-            SharedDynamicArchivePath = const_cast<char *>(SharedArchiveFile);
-            SharedArchivePath = get_default_shared_archive_path();
             DynamicDumpSharedSpaces = true;
+            ArchiveClassesAtExit = const_cast<char *>(SharedArchiveFile);
+            SharedArchivePath = get_default_shared_archive_path();
```
Comment (iklam) [inline src/hotspot/share/runtime/arguments.cpp:3569]: I think we should set SharedArchiveFile=NULL here. That way, when foo.jsa doesn't exist, the two command lines are equivalent:

- java -XX:ArchiveClassesAtExit=foo.jsa ...
- java -XX:SharedArchiveFile=foo.jsa -XX:+AutoCreateSharedArchive ...

and we can revert the changes on line 3602, to make the code easier to understand.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java [hunk 8]
```diff
@@ -0,0 +1,693 @@
...
+ *  (A) Test with default base archive -XX:+SharedArchiveFile=<archive>
+ *  (B) Test with the base archive is specified: -XX:SharedArchiveFile=<base>:<file>
+ *  all the following if not explained explicitly, run with flag -XX:+AutoCreateSharedArchive
+ *
+ *  Note VerifySharedSpaces will affect output so the tests run twice: one with -XX:+VerifySharedSpaces and the other with -XX:-VerifySharedSpaces
+ *
+ * 10 Case (A)
+ *
+ *   10.01 run with non-existing archive should automatically create dynamic archive.
+ *        If the JDK's default CDS archive cannot be loaded, print out warning, run continue without shared archive and no shared archive created at exit.
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java:51]: The test cases are called "10.01" here but "10.1" before. This makes it hard to do a text search. Could you make the numbering consistent?

**Review by iklam — COMMENTED** [review]
Some test comments

File: src/hotspot/share/runtime/java.cpp [hunk 9]
```diff
@@ -504,9 +505,11 @@ void before_exit(JavaThread* thread) {
 
 #if INCLUDE_CDS
   if (DynamicArchive::should_dump_at_vm_exit()) {
-    assert(ArchiveClassesAtExit != NULL, "Must be already set");
+    const char* archive_name =
+       (ArchiveClassesAtExit != nullptr) ? ArchiveClassesAtExit : Arguments::GetSharedDynamicArchivePath();
+    assert(archive_name != NULL, "Must be already set");
```
Comment (iklam) [inline src/hotspot/share/runtime/java.cpp:510]: The above change doesn't seem necessary. `ArchiveClassesAtExit` should be already set to non-null if AutoCreateArchive was set but the dynamic archive cannot be mapped.

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 10]
```diff
@@ -946,12 +948,22 @@ void MetaspaceShared::initialize_runtime_shared_and_meta_spaces() {
...
     }
   } else {
     set_shared_metaspace_range(NULL, NULL, NULL);
     UseSharedSpaces = false;
+    if (AutoCreateSharedArchive) {
+      AutoCreateSharedArchive = false;
+    }
+    if (DynamicDumpSharedSpaces) {
+      DynamicDumpSharedSpaces = false;
+    }
```
Comment (iklam) [inline src/hotspot/share/cds/metaspaceShared.cpp:966]: I think there's no need to put the above checks in 'if' blocks. I would suggest replacing the above six lines with:
```
// The base archive cannot be mapped. We cannot dump the dynamic shared archive.
AutoCreateSharedArchive = false;
DynamicDumpSharedSpaces = false;
```

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 11]
```diff
@@ -946,12 +948,22 @@ void MetaspaceShared::initialize_runtime_shared_and_meta_spaces() {
     _requested_base_address = static_mapinfo->requested_base_address();
     if (dynamic_mapped) {
       FileMapInfo::set_shared_path_table(dynamic_mapinfo);
+      // turn AutoCreateSharedArchive off if successfully mapped
+      if (AutoCreateSharedArchive) {
+        AutoCreateSharedArchive = false;
+      }
```
Comment (iklam) [inline src/hotspot/share/cds/metaspaceShared.cpp:954]: No need for the 'if' check.

File: src/hotspot/share/cds/filemap.cpp [hunk 12]
```diff
@@ -1234,15 +1229,18 @@ bool FileMapInfo::check_archive(const char* archive_name, bool is_static) {
...
 
-  if (!file_helper.initialize(archive_name)) {
+  if (!file_helper.initialize()) {
     return false;
   }
   GenericCDSFileMapHeader* header = file_helper.get_generic_file_header();
   if (header->_magic != CDS_DYNAMIC_ARCHIVE_MAGIC) {
     assert(header->_magic == CDS_ARCHIVE_MAGIC, "must be");
+    if (AutoCreateSharedArchive) {
+      warning("AutoCreateSharedArchive is ignored because %s is a static archive", archive_name);
```
Comment (iklam) [inline src/hotspot/share/cds/filemap.cpp:1242]: We should use log_warning(cds) for all new warning messages.

**Review by iklam — COMMENTED** [review]
I have reviewed all the files. The changes look good overall. Just a few small nits.

File: src/hotspot/share/runtime/java.cpp [hunk 13]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1997, 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (calvinccheung) [inline src/hotspot/share/runtime/java.cpp:2]: You can revert the copyright header change since there's no other changes in this file.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchiveNoDefaultArchive.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+                      .shouldContain("UseSharedSpaces: Initialize static archive failed")
+                      .shouldContain("UseSharedSpaces: Unable to map shared spaces")
+                      .shouldContain("mixed mode")
+                      .shouldNotContain("sharing");
+        }
+        // delete existing jsa file
+        if (jsaFile.exists()) {
+            jsaFile.delete();
+        }
+        System.out.println("======= run with no default shared archive should no create shared archive at exit");
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchiveNoDefaultArchive.java:126]: Typo: should no create -> should not create

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java [hunk 15]
```diff
@@ -0,0 +1,693 @@
...
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller -jar WhiteBox.jar sun.hotspot.WhiteBox
+ * @run main/othervm -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xbootclasspath/a:./WhiteBox.jar TestAutoCreateSharedArchive verifySharedSpacesOff
+ * @run main/othervm -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xbootclasspath/a:./WhiteBox.jar TestAutoCreateSharedArchive verifySharedSpacesOn
+ */
+
+/*
+ * -XX:SharedArchiveFile can be specified in two styles:
+ *
+ *  (A) Test with default base archive -XX:+SharedArchiveFile=<archive>
+ *  (B) Test with the base archive is specified: -XX:SharedArchiveFile=<base>:<top>
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java:43]: The "is" is extra.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java [hunk 16]
```diff
@@ -0,0 +1,693 @@
...
+ *   13.01  run with a bad versioned (< CDS_GENERIC_HEADER_SUPPORTED_MIN_VERSION) archive should not create dynamic archive at exit.
+ *   13.02  run with a bad versioned (> CDS_GENERIC_HEADER_SUPPORTED_MIN_VERSION) archive should create dynamic archive at exit.
+ *
+ * 14 run with an archive whose base name is not matched, no shared archive at exit.
+ *
+ * 15 run with an archive whose jvm_ident is corrupted should
+ *     create dynamic archive at exit with -XX:-VerifySharedSpaces
+ *     not create dynamic archive at exit with -XX:+VerifySharedSpaces
+ *
+ * 16 run with an archive of only containing magic in the file (size of 4 bytes)
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java:72]: Please delete the "of" after "archive".

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java [hunk 17]
```diff
@@ -0,0 +1,693 @@
...
+ *
+ *   20.01 dump base archive which will be used for dumping top archive.
+ *   20.02 dump top archive based on base archive obtained in 20.1.
+ *   20.03 run -XX:SharedArchiveFile=<base>:<top> to verify the archives.
+ *   20.04 run with -XX:SharedArchveFile=base:top (reversed)
+ *
+ * 21 Mismatched versions
+ *   21.01 if version of top archive is higher than CDS_GENERIC_HEADER_SUPPORTED_MIN_VERSION, the archive cannot be shared and will be
+ *         regenerated at exit.
+ *   21.02 if version of top archive is lower than CDS_GENERIC_HEADER_SUPPORTED_MIN_VERSION, the archive cannot be shared abd will be
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java:85]: s/abd/and

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java [hunk 18]
```diff
@@ -0,0 +1,693 @@
...
+
+    public static void print(String message) {
+        System.out.println(message);
+    }
+
+    private static void testAutoCreateSharedArchive() throws Exception {
+        String appJar = ClassFileInstaller.getJarPath("hello.jar");
+        boolean fileModified = false;
+
+        String VerifySharedSpaces = verifyOn ? "-XX:+VerifySharedSpaces" : "-XX:-VerifySharedSpaces";
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java:154]: A local variable name usually starts with lowercase.
Suggest changing VerifySharedSpaces to verifySharedSpaces.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java [hunk 19]
```diff
@@ -0,0 +1,693 @@
...
+            mainAppClass)
+            .assertNormalExit(output -> {
+                output.shouldHaveExitValue(0)
+                      .shouldContain(HELLO_WORLD)
+                      .shouldNotContain("Dumping shared data to file");
+                });
+        ft2 = Files.getLastModifiedTime(Paths.get(modMagic));
+        fileModified = !ft1.equals(ft2);
+        if (fileModified) {
+            throw new RuntimeException("Shared archive " + modMagic + " should automatically be generated");
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java:279]: Should the message be: “… should not be automatically generated”?

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java [hunk 20]
```diff
@@ -0,0 +1,693 @@
...
+             mainAppClass)
+             .assertNormalExit(output -> {
+                 output.shouldContain(HELLO_WORLD)
+                       .shouldContain("Unable to read generic CDS file map header from shared archive")
+                       .shouldContain("Dumping shared data to file:");
+             });
+        ft2 = Files.getLastModifiedTime(Paths.get(magicOnly));
+        fileModified = !ft1.equals(ft2);
+        if (!fileModified) {
+            throw new RuntimeException("Shared archive " + magicOnly + " should not be created at exit");
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java:578]: Should the message be: “… should be created at exit”?

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java [hunk 21]
```diff
@@ -0,0 +1,693 @@
...
+             mainAppClass)
+             .assertNormalExit(output -> {
+                 output.shouldContain("Specified shared archive not found (" + nonExistBase + ")")
+                       .shouldContain(HELLO_WORLD)
+                       .shouldNotContain("Dumping shared data to file:");
+             });
+        ft2 = Files.getLastModifiedTime(Paths.get(TOP_NAME));
+        fileModified = !ft1.equals(ft2);
+        if (fileModified) {
+            throw new RuntimeException("Shared archive " + TOP_NAME + " should be created at exit");
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchive.java:690]: Should the message be: “…should not be created at exit”?

**Review by calvinccheung — COMMENTED** [review]
Looks good. Some minor nits.

**Review by iklam — APPROVED** [review]
LGTM

**Review by calvinccheung — APPROVED** [review]
LGTM

**Comment by yminqi** [comment]
@iklam @calvinccheung 
Thanks for review!
